src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -57,6 +63,89 @@ def _build_x509_name(backend, x509_name):
     return x509.Name(attributes)
 
 
+def _build_general_name(backend, gn):
+    # dNSName
+    if gn.type == 2:
+        data = backend._ffi.buffer(
+            gn.d.dNSName.data, gn.d.dNSName.length
+        )[:].decode("ascii")
+        value = x509.DNSName(idna.decode(data))
+    # dirName
+    elif gn.type == 4:
+        value = x509.DirectoryName(
+            _build_x509_name(backend, gn.d.directoryName)
+        )
+    # uniformResourceIdentifier
+    elif gn.type == 6:
+        data = backend._ffi.buffer(
+            gn.d.uniformResourceIdentifier.data,
+            gn.d.uniformResourceIdentifier.length
+        )[:].decode("ascii")
+        parsed = urllib_parse.urlparse(data)
+        hostname = idna.decode(parsed.hostname)
+        if parsed.port:
+            netloc = hostname + u":" + str(parsed.port)
+        else:
+            netloc = hostname
+
+        # Note that building a URL in this fashion means it should be
+        # semantically indistinguishable from the original but is not
+        # guaranteed to be exactly the same.
+        uri = urllib_parse.urlunparse((
+            parsed.scheme,
+            netloc,
+            parsed.path,
+            parsed.params,
+            parsed.query,
+            parsed.fragment
+        ))
+        value = x509.UniformResourceIdentifier(uri)
+    # rfc822Name
+    elif gn.type == 1:
+        data = backend._ffi.buffer(
+            gn.d.rfc822Name.data, gn.d.rfc822Name.length
+        )[:].decode("ascii")
+        parsed = parseaddr(data)
+        if parsed[1]:
+            parts = parsed[1].split("@")
+            assert len(parts) <= 2
+            if len(parts) == 1:
+                value = x509.RFC822Name(data)
+            elif parsed[0]:
+                value = x509.RFC822Name(
+                    parsed[0] + u" <" + parts[0] +
+                    u"@" + idna.decode(parts[1]) + u">"
+                )
+            else:
+                value = x509.RFC822Name(
+                    parts[0] + u"@" + idna.decode(parts[1])
+                )
+        else:
+            value = x509.RFC822Name(data)
+    # iPAddress
+    elif gn.type == 7:
+        value = x509.IPAddress(
+            ipaddress.ip_address(
+                backend._ffi.buffer(
+                    gn.d.iPAddress.data, gn.d.iPAddress.length
+                )[:]
+            )
+        )
+    # registeredID
+    elif gn.type == 8:
+        oid = _obj2txt(backend, gn.d.registeredID)
+        value = x509.RegisteredID(x509.ObjectIdentifier(oid))
+    else:
+        # otherName, x400Address or ediPartyName
+        raise x509.UnsupportedGeneralNameType(
+            "{0} is not a supported type".format(
+                x509._GENERAL_NAMES[gn.type]
```

Comment:
I was thinking just drop the text, number only. Simplicity or something.

On Mon, Apr 20, 2015 at 10:13 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1855#discussion_r28745483>:
>
> > +        value = x509.IPAddress(
> > +            ipaddress.ip_address(
> > +                backend._ffi.buffer(
> > +                    gn.d.iPAddress.data, gn.d.iPAddress.length
> > +                )[:]
> > +            )
> > +        )
> > +    # registeredID
> > +    elif gn.type == 8:
> > +        oid = _obj2txt(backend, gn.d.registeredID)
> > +        value = x509.RegisteredID(x509.ObjectIdentifier(oid))
> > +    else:
> > +        # otherName, x400Address or ediPartyName
> > +        raise x509.UnsupportedGeneralNameType(
> > +            "{0} is not a supported type".format(
> > +                x509._GENERAL_NAMES[gn.type]
>
> Probably not. I could do .get on this to handle that case and just have it
> return the number then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1855/files#r28745483>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084